src/appmixer/google/gmail/NewAttachment/NewAttachment.js

```diff
@@ -1,96 +1,55 @@
-        });
-    });
-};
-
-/**
- * Component will trigger anytime there is a new email with attachment(s) in your mailbox
- * and send those attachments to output port.
- * @extends {Component}
- */
 module.exports = {
-
     async tick(context) {
-
-        let auth = commons.getOauth2Client(context.auth);
         let newState = {};
+        let auth = commons.getOauth2Client(context.auth);
+        let { userId } = context.auth;
 
-        let result = await commons.listNewMessages(
-            { auth: auth, userId: 'me', quotaUser: context.auth.userId },
+        const { labels: { AND: labels } = { AND: [] } } = context.properties;
+        const isLabelsEmpty = !labels.some(label => label.name);
+
+        const data = await commons.listNewMessages({ auth, userId: 'me', quotaUser: userId },
             context.state.id || null);
 
         // latest message or there are no messages in the inbox
-        newState.id = result.lastMessageId;
+        newState.id = data.lastMessageId;
 
-        let emails = await Promise.map(result.newMessages, async diff => {
-            return getMessage({
-                auth,
-                userId: 'me',
-                quotaUser: context.auth.userId,
-                format: 'full',
-                id: diff.id
-            }).catch(err => {
+        const emails = await Promise.map(data.newMessages, async message => {
+            return emailCommons.callEndpoint(context, `/users/me/messages/${message.id}`, {
+                method: 'GET',
+                params: { format: 'full' }
+            }).then(response => response.data).catch(err => {
                 // email can be deleted (permanently) in gmail between listNewMessages call and
                 // this getMessage call, in such case - ignore it and return null
-                if (err && err.code === 404) {
+                if (err && err.response && err.response.status === 404) {
                     return null;
                 }
                 throw err;
             });
-        });
+        }, { concurrency: 10 });
 
         let attachments = await Promise.map(emails, email => {
-            if (!email) {
-                // this is the case, when new email was deleted before we could get it's details in previous step
+            if (!email || !email.labelIds) {
+                // this is the case, when new email was deleted before we could get its details in the previous step
                 return [];
             }
-            return downloadAttachments(auth, context, email);
+            // Filter emails based on selected labels
+            if (isLabelsEmpty || labels.some(label => email.labelIds.includes(label.name))) {
+                return downloadAttachments(auth, context, email);
+            }
+            return [];
         });
 
         attachments = attachments.reduce((a, b) => a.concat(b), []);
 
         let saved = await Promise.map(attachments, attachment => {
             return context.saveFile(
```

Comment:
context.saveFile  is deprecated, use the `context.saveFilestream` instead. 
also please use the streams as described in the comment for the `GetAttachment`

Reply:
replaced with saveFilestream